### PR TITLE
Cypress: Initial addition of target based port of mcuboot

### PR DESCRIPTION
**Pull request by romanjoe**
This PR contains all the code needed to compile and run mcuboot based bootloader on cy8cproto_062_4343w development board.

All the code is contained in  boot/cypress folder and do not change any mcuboot source code. Makefile based build support implemented.

Submodules needed to be downloaded for successful build.

There is also a `BlinkyApp` - test application to quickly test `MCUBootApp` which is a bootloader.

Information regarding bootloader and test application build can be found in corresponding Readme.md files.

**Note:**
one of submodules boot/cypress/libs/bsp/TARGET_CY8CPROTO-062-4343W has reference to private cypress repo and will be updated to once available in public repo.

Any comments are welcome.

### CODE REVIEW

File: boot/cypress/MCUBootApp/keys.c [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+ /*******************************************************************************
+* \copyright
+*
+* (c) 2018, Cypress Semiconductor Corporation
```
Comment (d3zd3z) [inline boot/cypress/MCUBootApp/keys.c]: I'm not sure if this license is compatible with the Apache-2.0 license above this, but Apache can be tricky to apply. This particular license does restrict the content of this file to use on Cypress devices, so we would need to get general agreement by all of the users of MCUboot to be able to put this in the main repo. Given that the file doesn't appear to have any copyrightable content, perhaps the Cypress-specific license could be removed?

File: boot/cypress/MCUBootApp/libs.mk [hunk 2]
```diff
@@ -0,0 +1,90 @@
+################################################################################
```
Comment (d3zd3z) [inline boot/cypress/MCUBootApp/libs.mk]: This file has DOS line endings. It should be Unixy endings to be consistent with the rest of the repo.

File: boot/cypress/MCUBootApp/os/os.h [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+* \file os.h
+* \version 1.0
+*
+* \brief
+* Objective:
+*    Heap init prototype
+*
+********************************************************************************
+* \copyright
+* Copyright 2017-2018, Cypress Semiconductor Corporation. All rights reserved.
```
Comment (d3zd3z) [inline boot/cypress/MCUBootApp/os/os.h]: Can we just use Apache for this?

File: boot/cypress/MCUBootApp/sysflash/sysflash.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+#define FLASH_DEVICE_INTERNAL_FLASH        (0x7F)
+
+#define FLASH_AREA_BOOTLOADER 0
+#define FLASH_AREA_IMAGE_0 1
+#define FLASH_AREA_IMAGE_1 2
+#define FLASH_AREA_IMAGE_SCRATCH 3
+#define FLASH_AREA_IMAGE_2 5
+#define FLASH_AREA_IMAGE_3 6
+
+#define CY_FLASH_ALIGN                      (128) // TODO: will fix it to CY_FLASH_SIZEOF_ROW (512) in PSoC6
```
Comment (d3zd3z) [inline boot/cypress/MCUBootApp/sysflash/sysflash.h]: I think now, you should be able to make the flash align function return a larger integer type, and make this the correct value.

File: boot/cypress/README.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+* BlinkyApp - simple PSoC6 blinking LED application which is a target of BOOT/UPGRADE;
+
+The device supported is CY8CPROTO-062-4343W board which is PSoC6 device with 2M of Flash available.
+
+The default flash map implemented is next:
+
+* [0x10000000, 0x10020000] - MCUBootApp (bootloader) area;
+* [0x10020000, 0x10030000] - primary slot for BlinkyApp;
+* [0x10030000, 0x10040000] - secondary slot for BlinkyApp;
+* [0x10030000, 0x10040000] - secondary slot for BlinkyApp;
```
Comment (d3zd3z) [inline boot/cypress/README.md]: Duplicate line?

File: boot/cypress/README.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+Those are represented as submodules.
+
+To retrieve source code with subsequent submodules pull:
+
+    git clone --recursive http://git-ore.aus.cypress.com/repo/cy_mcuboot_project/cy_mcuboot.git
+
+Submodules can also be updated and initialized separately:
+
+    cd cy_mcuboot
+    git submodule update --init
```
Comment (d3zd3z) [inline boot/cypress/README.md]: You can add `--recursive` to the submodule update, and won't need the second below on pulling in the crypto library.

File: boot/cypress/cy_flash_pal/cy_flash_psoc6.c [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+* \version 1.0
+*
+* \brief
+*  This is the source file of flash driver adoption layer between PSoC6 
+*  and standard MCUBoot code.
+*
+********************************************************************************
+* \copyright
+*
+* (c) 2019, Cypress Semiconductor Corporation
```
Comment (d3zd3z) [inline boot/cypress/cy_flash_pal/cy_flash_psoc6.c]: Can this be made Apache-2.0?

File: boot/cypress/cy_flash_pal/cy_flash_map.c [hunk 8]
```diff
@@ -0,0 +1,494 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/***************************************************************************//**
+* \file cy_flash_pal.c
```
Comment (d3zd3z) [inline boot/cypress/cy_flash_pal/cy_flash_map.c]: This name I'm guessing is a file this was derived from.

File: boot/cypress/cy_flash_pal/cy_flash_map.c [hunk 9]
```diff
@@ -0,0 +1,494 @@
...
+*
+* \brief
+*  This is the source file of flash driver adaptation layer between PSoC6
+*  and standard MCUBoot code.
+*
+********************************************************************************
+* \copyright
+*
+* (c) 2019, Cypress Semiconductor Corporation
+* or a subsidiary of Cypress Semiconductor Corporation. All rights
```
Comment (d3zd3z) [inline boot/cypress/cy_flash_pal/cy_flash_map.c]: Again, it'd be nice if we could just make this Apache-2.0.

File: boot/cypress/cy_flash_pal/cy_flash_map.c [hunk 10]
```diff
@@ -0,0 +1,494 @@
...
+#ifndef CY_BOOT_EXTERNAL_FLASH_ERASE_VALUE
+/* This is the value of external flash bytes after an erase */
+#define CY_BOOT_EXTERNAL_FLASH_ERASE_VALUE      (0xff)
+#endif
+
+#ifdef CY_FLASH_MAP_EXT_DESC
+/* Nothing to be there when external FlashMap Descriptors are used */
+#else
+static struct flash_area bootloader =
+{
```
Comment (d3zd3z) [inline boot/cypress/cy_flash_pal/cy_flash_map.c]: Typically, our convention puts the brace on the same line as the equal sign, reserving a brace at the beginning of a line for function declarations.

File: boot/cypress/cy_flash_pal/cy_flash_map.c [hunk 11]
```diff
@@ -0,0 +1,494 @@
...
+};
+#endif
+
+/*< Opens the area for use. id is one of the `fa_id`s */
+int flash_area_open(uint8_t id, const struct flash_area **fa)
+{
+    int ret = -1;
+    uint32_t i = 0;
+
+    while(NULL != boot_area_descs[i])
```
Comment (d3zd3z) [inline boot/cypress/cy_flash_pal/cy_flash_map.c]: Should be a space between the `while` and the open paren.

File: boot/cypress/cy_flash_pal/cy_flash_map.c [hunk 12]
```diff
@@ -0,0 +1,494 @@
...
+    {
+        if(id == boot_area_descs[i]->fa_id)
```
Comment (d3zd3z) [inline boot/cypress/cy_flash_pal/cy_flash_map.c]: Likewise, a space after the if. It also makes it consistent with the rest of the file.

File: boot/cypress/cy_flash_pal/cy_flash_map.c [hunk 13]
```diff
@@ -0,0 +1,494 @@
...
+#endif
+
+/*< Opens the area for use. id is one of the `fa_id`s */
+int flash_area_open(uint8_t id, const struct flash_area **fa)
+{
+    int ret = -1;
+    uint32_t i = 0;
+
+    while(NULL != boot_area_descs[i])
+    {
```
Comment (d3zd3z) [inline boot/cypress/cy_flash_pal/cy_flash_map.c]: At least the rest of the mcuboot convention places this brace at the end of the previous line.

File: boot/cypress/cy_flash_pal/cy_flash_psoc6.c [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+	while(rowNum>0)
+	{
+		rc = Cy_Flash_EraseRow(address);
+		assert(rc == 0);
+		address += CY_FLASH_SIZEOF_ROW;
+		rowNum--;
+	}
+
+	/* if Start of erase area is unaligned */
+	if(remStart != 0)
```
Comment (d3zd3z) [inline boot/cypress/cy_flash_pal/cy_flash_psoc6.c]: This really should just be an error. MCUboot should never be erasing regions that are misaligned, and by reading them, erasing, and then writing, you create a window where an untimely powerdown will result in a loss of data.

If you are finding the code calling with misaligned erases, we need to figure out what is happening and fix that.

File: boot/cypress/cy_flash_pal/cy_flash_psoc6.c [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+* \return
+*  PSOC6_WR_SUCCESS                     A successful write
+*  PSOC6_WR_ERROR_INVALID_PARAMETER     At least one of the input parameters is invalid
+*  PSOC6_WR__ERROR_FLASH_WRITE          Error in flash Write
+*
+*******************************************************************************/
+int psoc6_flash_write_hal(uint8_t data[],
+							 uint32_t address,
+							 uint32_t len)
+{
```
Comment (d3zd3z) [inline boot/cypress/cy_flash_pal/cy_flash_psoc6.c]: This also doesn't need to be this complex. MCUboot should never try writing to an area of flash that hasn't been erased. If you find that happening, it should cause a failure which can be investigated.

File: boot/cypress/cy_flash_pal/cy_flash_psoc6.c [hunk 16]
```diff
@@ -0,0 +1,277 @@
...
+        while((srcIndex < len) && (rc == CY_FLASH_DRV_SUCCESS))
+        {
+            rowsNotEqual = 0u;
+            /* Copy data to the write buffer either from the source buffer or from the flash */
+            for(dstIndex = 0u; dstIndex < CY_FLASH_SIZEOF_ROW; dstIndex++)
+            {
+                if((byteOffset >= eeOffset) && (srcIndex < len))
+                {
+                    writeBufferPointer[dstIndex] = data[srcIndex];
+                    /* Detect that row programming is required */
```
Comment (d3zd3z) [inline boot/cypress/cy_flash_pal/cy_flash_psoc6.c]: This shouldn't ever happen, and really should just be an error.

File: boot/cypress/cy_flash_pal/include/cy_flash_psoc6.h [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+* \file cy_flash_psoc6.h
+* \version 1.0
+*
+* \brief
+*  This is the header file for PSoC6 flash driver adoption layer.
+*
+********************************************************************************
+* \copyright
+*
+* © 2019, Cypress Semiconductor Corporation
```
Comment (d3zd3z) [inline boot/cypress/cy_flash_pal/include/cy_flash_psoc6.h]: Apache?

File: boot/cypress/cy_flash_pal/include/flash_map_backend/flash_map_backend.h [hunk 18]
```diff
@@ -0,0 +1,247 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+ /*******************************************************************************
+* \copyright
+*
+* (c) 2019, Cypress Semiconductor Corporation
```
Comment (d3zd3z) [inline boot/cypress/cy_flash_pal/include/flash_map_backend/flash_map_backend.h]: Same license issue as elsewhere.

File: boot/cypress/keys/cypress-test-ec-p256.pem [hunk 19]
```diff
@@ -0,0 +1,5 @@
+-----BEGIN PRIVATE KEY-----
```
Comment (d3zd3z) [inline boot/cypress/keys/cypress-test-ec-p256.pem]: Any reason to make these keys specific to the cypress port, and not just use the keys at the root of the repo (which are intended for testing)?

**Comment by utzig** [comment]
@romanjoe I am trying to review on a commit by commit basis, but after the first commit (ce613b8), the following few commits I have looked into (7960931, 44e8d63, cb11c83) already basically fix issues from the initial commit (apart from adding a CI config file). Wouldn't it be better to actually squash commits in a way that the final work consists only of finished and known to work changes? It could be as many commits as you wish as long as they don't keep piling up fixes!

File: boot/cypress/keys/cypress-test-ec-p256.pem (hunk 19 above)
Comment (BohdanCy) [reply boot/cypress/keys/cypress-test-ec-p256.pem]: We supplied these key the same as are supplied within cysecuretools, so signed images will be consistent for al Cy-related Secure Boot FW.

**Comment by d3zd3z** [comment]
Where are we at on this? Last I'd heard we were waiting for one of the depend repositories to be updated. This will definitely need to be remade into a small number of commits to make it reviewable.

File: boot/cypress/cy_flash_pal/cy_flash_psoc6.c (hunk 15 above)
Comment (romanjoe) [reply boot/cypress/cy_flash_pal/cy_flash_psoc6.c]: We have opened a jira to rework this API and plan to submit this via another PR

**Comment by d3zd3z** [comment]
I think at this point, this code is going to be easier to review if we make fresh commits (or at least squash). I'd guess maybe just a handful of commits adding the new functionality.

**Comment by romanjoe** [comment]
Hi David,

Last thing i did here was rearranging of commits and squashing them into a more meaningful history after chatting with Fabio, then i did fixes per your comments as separate commit. Do you suggest now to shrink current history into some couple commits, kinda "Added initial support fo 062 Cypress devices"?

Thank you

**Comment by utzig** [comment]
@romanjoe Please also rebase over latest `master`.

**Comment by d3zd3z** [comment]
> "Added initial support fo 062 Cypress devices"

Except, please try to write commit messages in a simple present tense. Writing commit text in the past tense makes them slightly more awkward to read when later looking through commits. It isn't all that big of a deal, but it does help a little bit later on. So, "Add initial support for 062 Cypress devices" would be fine, at least as far as tense goes.

As far as updates and squashing, it is usually helpful to make separate commits to address issues brought up by reviews. It makes it easier to see that they've been addressed. But, keep them associated with existing commits so that it will be easier to squash before we merge everything.

**Comment by d3zd3z** [comment]
Let me try to clarify a little bit about writing commit text in the present tense.

When writing a commit, it kind of makes sense to refer to the changes in the past tense, since it is describing something that you did before writing that message. However, the commit text is part of that change, and from the perspective of the commit text, it makes more sense to refer to the changes more actively, and in present tense. This becomes more important with git, because commits can be manipulated (cherry-picked, etc).

It is also important when commit text needs to refer to other things that have happened in the tree. It makes sense to refer to earlier commits in the past tense (and occasionally to refer to upcoming commits in the future).

Think of the commit text as describing what that commit is doing to the code. "Add foo", "Replace bar", etc.

**Comment by utzig** [comment]
@romanjoe This PR can closed, right?

**Comment by romanjoe** [comment]
@utzig yes, thank you

**Pull request closed without merging**